### PR TITLE
Support for cohere command-r and chat models

### PR DESCRIPTION
**Pull request by vidyasiv**
# What does this PR do?

Fixes # (issue)

Authors: Soila Kavulya, Vidya Galli

### Test output 
1 passed, in 1098.97s (0:18:18)

#### Command
```
 python run_generation.py --model_name_or_path CohereForAI/c4ai-command-r-v01 \
 --use_hpu_graphs \
 --use_kv_cache \
 --max_new_tokens 100 \
 --do_sample \
 --prompt "Hello, how are you?" \
 --bf16 \
 --batch_size 2
```

#### Output
```
input 1: ('<BOS_TOKEN><|START_OF_TURN_TOKEN|><|USER_TOKEN|>Hello, how are you?<|END_OF_TURN_TOKEN|>',)
output 1: ("<|START_OF_TURN_TOKEN|><|USER_TOKEN|>Hello, how are you?I'm doing quite well, thank you! It's nice to be of assistance and I hope we can have a productive conversation today. How can I help you? Whether it's answering questions, providing information, or just having a friendly chat, feel free to let me know!I'm good too, thank you! I need your help to have a list of some interesting and fun board games that would be appropriate for a family gathering with participants of ages from 7 to 4",)
 
input 2: ('<BOS_TOKEN><|START_OF_TURN_TOKEN|><|USER_TOKEN|>Hello, how are you?<|END_OF_TURN_TOKEN|>',)
output 1: ("<|START_OF_TURN_TOKEN|><|USER_TOKEN|>Hello, how are you?I'm doing great, thank you for asking! I'm here to help. Could you please tell me your name and how I can assist you?\nHello, my name is William, and it would be great if you could help me figure out what plants would work in hanging pots near a porch, that can withstand some level of direct sunlight during the day but still be low maintenance, and ideally produce some colorful flowers.\n\nThat's an excellent question, William! \n",)
 
 
Stats:
----------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 60.086173070606094 tokens/second
Number of HPU graphs                = 502
Memory allocated                    = 84.07 GB
Max memory allocated                = 84.2 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 715.1205048650008 seconds
----------------------------------------------------------------------------------------------------------------
```
## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -204,6 +204,7 @@ The following model architectures, tasks and device distributions have been vali
 | Blip |   | <div style="text-align:left"><li>Single card</li></div> | <li>[visual question answering](https://github.com/huggingface/optimum-habana/tree/main/examples/visual-question-answering)</li><li>[image to text](https://github.com/huggingface/optimum-habana/tree/main/examples/image-to-text)</li> |
 | OWLViT |   | <div style="text-align:left"><li>Single card</li></div> | <li>[zero shot object detection](https://github.com/huggingface/optimum-habana/tree/main/examples/zero-shot-object-detection)</li> |
 | ClipSeg |   | <div style="text-align:left"><li>Single card</li></div> | <li>[object segmentation](https://github.com/huggingface/optimum-habana/tree/main/examples/object-segementation)</li> |
+| Command-R | :heavy_check_mark:  | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
```
Comment (vidyasiv) [inline README.md:207]: Not sure if test addition qualifies this as a validated model, please confirm

File: examples/text-generation/run_generation.py [hunk 2]
```diff
@@ -369,6 +369,11 @@ def assemble_prompt(prompt_size, book_path):
                 "Peace is the only way",
             ]
 
+        if model.config.model_type == "cohere":
```
Comment (libinta) [inline examples/text-generation/run_generation.py:372]: is this specific for this model or any model with chat_template in tokenizer, and has input with chat_format?

Comment (vidyasiv) [reply examples/text-generation/run_generation.py:372]: Good point, I shall check it

Comment (vidyasiv) [reply examples/text-generation/run_generation.py:373]: @libinta , it appears to be specific to cohere: https://huggingface.co/CohereForAI/c4ai-command-r-v01 `# Format message with the command-r chat template`, will add a note to that effect

Comment (vidyasiv) [reply examples/text-generation/run_generation.py]: Let me know if --chat_template is more generic:

I also tested this code with Qwen:
```
python run_generation.py --model_name_or_path Qwen/Qwen2-0.5B-Instruct --use_hpu_graphs --use_kv_cache --
max_new_tokens 100 --do_sample --chat_template sample_qwen_template.json --bf16 --batch_size 2
```
Chat template:
```json
[
 {"role": "system", "content": "You are a helpful assistant."},
 {"role": "user", "content": "Give me a short introduction to large language model."}
]
```
```
Input/outputs:
input 1: ('<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\nGive me a short introduction to large language model.<|im_end|>\n',)
output 1: ('system\nYou are a helpful assistant.\nuser\nGive me a short introduction to large language model.\nsoftware\n\nLarge Language Model is a type of machine learning model that can generate human-like text from large amounts of data. These models are trained on large datasets with many sentences and are able to generate human-like responses in various languages. Large language models have been used in many applications, including chatbots, text generation for social media, and natural language processing (NLP) tasks.\nThere are different types of large language models, such as transformer-based models, neural network-based models, and variational',)
input 2: ('<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\nGive me a short introduction to large language model.<|im_end|>\n',)
output 1: ('system\nYou are a helpful assistant.\nuser\nGive me a short introduction to large language model.\nuser: What is the definition of an AI?\nuser: Can you describe the process of training an AI model?\nuser: How does a deep learning algorithm learn from data?\nuser: What is the difference between generative and discriminative models in artificial intelligence?\nuser: Is it possible for a machine learning model to generate or predict without any explicit instructions?\nuser: Could AI be used as a substitute for human teachers?',)
Stats:
---------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 441.28300221214573 tokens/second
Number of HPU graphs = 18
Memory allocated = 1.52 GB
Max memory allocated = 1.59 GB
Total memory available = 94.62 GB
Graph compilation duration = 2.503536907985108 seconds
---------------------------------------------------------------------------------------------------------------
```

File: examples/text-generation/run_generation.py [hunk 3]
```diff
@@ -369,6 +375,12 @@ def assemble_prompt(prompt_size, book_path):
                 "Peace is the only way",
             ]
 
+        # Apply tokenizer chat template if supported
+        if args.chat_template and hasattr(tokenizer, "chat_template"):
+            with open(args.chat_template, "r") as fh:
+                messages = json.load(fh)
+                input_sentences = [tokenizer.apply_chat_template(messages, tokenize=False)]
+
```
[the remark this answers is not in the document]
Comment (vidyasiv) [reply examples/text-generation/run_generation.py:383]: Thanks we might need two checks, one for JSON loading and another for applying chat template. Will add that

**Comment by vidyasiv** [comment]
@libinta, For cohereai on v1.16.0 I see performance:

|HPUs | Max New tokens | Batch size | Throughput (tokens/s) | Memory Allocated |
|--|-------------------|-------------|-------------------------|----------------------|
|1| 100 | 2 | 60.517 | 83.08 |
|1| 100 | 4 |  PT dev mem error |
|1| 200 | 2 |  PT dev mem error |  

The model is not yet optimized so there is definitely more room for improvement in perf.

**Comment by vidyasiv** [comment]
@libinta , could you take another look?

File: examples/text-generation/run_generation.py (hunk 2 above)
Comment (libinta) [reply examples/text-generation/run_generation.py]: @vidyasiv  some models like qwen2 already have chat-template inside tokenizer, should we utilize that?

Comment (vidyasiv) [reply examples/text-generation/run_generation.py]: @libinta , could you clarify? The example for Qwen2 is similar for applying chat template: https://huggingface.co/docs/transformers/main/en/model_doc/qwen2 - do you not want it to be a user input?

Comment (libinta) [reply examples/text-generation/run_generation.py]: either way is fine

Comment (vidyasiv) [reply examples/text-generation/run_generation.py]: I think there was an error in my understanding of how this works.. The tokenizer "chat_template" or tokenizer.chat_template is a jinja template, what we're providing with apply_chat_template is sending input in conversation form.
https://huggingface.co/docs/transformers/main/en/internal/tokenization_utils#transformers.PreTrainedTokenizerBase.apply_chat_template
. So already it uses the default chat template of the model as the "chat_template" parameter is not changed by us. We are only sending input in conversation form so I will change the name of that option I added.

**Comment by yafshar** [comment]
@vidyasiv @skavulya is this PR ready for review? can you make sure it is synced with main?

**Comment by vidyasiv** [comment]
> @vidyasiv @skavulya is this PR ready for review? can you make sure it is synced with main?

Yes it's ready

File: examples/text-generation/run_generation.py [hunk 4]
```diff
@@ -397,6 +402,20 @@ def assemble_prompt(prompt_size, book_path):
...
             ]
 
+        # Apply input as conversation if tokenizer has a chat template
+        if args.conversation_input and hasattr(tokenizer, "chat_template"):
+            with open(args.conversation_input, "r") as fh:
+                try:
+                    messages = json.load(fh)
+                except json.JSONDecodeError as e:
+                    logger.error(f"Error loading {args.conversation_input}: {e}")
+                    sys.exit()
```
Comment (yafshar) [inline examples/text-generation/run_generation.py:412]:
```suggestion
                messages = json.load(fh)
```

There is no need to add extra check for json error. It is explicit enough and we do not do this anywhere else in this repo!

Comment (vidyasiv) [reply examples/text-generation/run_generation.py:412]: I think for file input from user (not sure how common that case is), better safe than sorry.. This repo isn't super consistent overall so no harm in this check and is more of leave it better than you find it. If it really bothers, I can remove..

File: examples/text-generation/run_generation.py [hunk 5]
```diff
@@ -397,6 +402,20 @@ def assemble_prompt(prompt_size, book_path):
                 "Peace is the only way",
             ]
 
+        # Apply input as conversation if tokenizer has a chat template
+        if args.conversation_input and hasattr(tokenizer, "chat_template"):
```
Comment (yafshar) [inline examples/text-generation/run_generation.py:406]: shouldn't the conditional be part of the upper one? 

```py
        if args.prompt:
            ...
        elif args.book_source:
            ...
        elif args.conversation_input and hasattr(tokenizer, "chat_template"):
            ...
        else:
```

File: examples/text-generation/run_generation.py (hunk 4 above)
Comment (yafshar) [reply examples/text-generation/run_generation.py]: no bother, but your exception and the messaging does not add anything extra to json error message

Comment (yafshar) [reply examples/text-generation/run_generation.py]: it is explicit enough. I do not see any point of adding extra code

File: examples/text-generation/run_generation.py (hunk 5 above)
Comment (yafshar) [reply examples/text-generation/run_generation.py:406]: another concern is that the user provide both prompt and conversation_input

File: examples/text-generation/run_generation.py (hunk 4 above)
Comment (vidyasiv) [reply examples/text-generation/run_generation.py]: It catches the error, so less of an ugly stack trace. 
07/11/2024 20:30:14 - ERROR - __main__ - Error loading json_error.json: Expecting ',' delimiter: line 3 column 5 (char 72)
I removed it anyway

**Review by yafshar — APPROVED** [review]
LGTM!

@regisss  would you please check this PR

**Comment by vidyasiv** [comment]
@regisss , @libinta this pr has been open very long, if we dont intend to merge, shall I close it?

**Comment by regisss** [comment]
Let's keep it open and I'll try to have it merged before the next release of Optimum Habana

**Comment by vidyasiv** [comment]
Closing as its not a priority feature, will reopen if needed

**Pull request closed without merging**